frontend/ui/widget/textboxwidget.lua

```diff
@@ -333,4 +361,121 @@ function TextBoxWidget:onHoldWord(callback, ges)
     return
 end
 
+-- Allow selection of one or more words (with no visual feedback)
+-- Gestures should be declared in widget using us (e.g dictquicklookup.lua)
+function TextBoxWidget:onHoldStartText(_, ges)
+    -- just store hold start position and timestamp, will be used on release
+    self.hold_start_x = ges.pos.x - self.dimen.x
+    self.hold_start_y = ges.pos.y - self.dimen.y
+    self.hold_start_tv = TimeVal.now()
+    return true
+end
+
+function TextBoxWidget:onHoldReleaseText(callback, ges)
+    if not callback then return end
+
+    local hold_end_x = ges.pos.x - self.dimen.x
+    local hold_end_y = ges.pos.y - self.dimen.y
+    local hold_duration = TimeVal.now() - self.hold_start_tv
+    hold_duration = hold_duration.sec + hold_duration.usec/1000000
+
+    -- Swap start and end if needed
+    local x0, y0, x1, y1
+    -- first, sort by y/line_num
+    local start_line_num = math.ceil(self.hold_start_y / self.line_height_px)
+    local end_line_num = math.ceil(hold_end_y / self.line_height_px)
+    if start_line_num < end_line_num then
+        x0, y0 = self.hold_start_x, self.hold_start_y
+        x1, y1 = hold_end_x, hold_end_y
+    elseif start_line_num > end_line_num then
+        x0, y0 = hold_end_x, hold_end_y
+        x1, y1 = self.hold_start_x, self.hold_start_y
+    else -- same line_num : sort by x
+        if self.hold_start_x <= hold_end_x then
+            x0, y0 = self.hold_start_x, self.hold_start_y
+            x1, y1 = hold_end_x, hold_end_y
+        else
+            x0, y0 = hold_end_x, hold_end_y
+            x1, y1 = self.hold_start_x, self.hold_start_y
+        end
+    end
+
+    -- similar code to find start or end is in _findWordEdge() helper
+    local sel_start_idx = self:_findWordEdge(x0, y0, "start")
+    local sel_end_idx = self:_findWordEdge(x1, y1, "end")
+
+    if not sel_start_idx or not sel_end_idx then
+        -- one or both hold points were out of text
+        return true
+    end
+
+    local selected_text = table.concat(self.charlist, "", sel_start_idx, sel_end_idx)
+    DEBUG("onHoldReleaseText (duration: "..hold_duration..") : "..
```

Comment:
Don't do string concatenation for debug call. The `DEBUG` call should be a no op for non-debug mode. String concatenation will create a new string which introduces runtime penalty. You can do `DEBUG("onHoldReleaseText (duration: ", hold_duration, "): ", sel_start_idx, ">", sel_end_idx, " = ", selected_text)` instead.